hparser.cpp

```diff
@@ -60,3 +61,366 @@ VariableExprNode* HParser::variable() {
+      match(decaf::token_type::ptLParen);
+      VariableExprNode* var_new = variable();
+      match(decaf::token_type::OpAssign);
+      ExprNode* value = expr_or();
+      match(decaf::token_type::ptSemicolon);
+      ExprNode* condition = expr_or();
+      match(decaf::token_type::ptSemicolon);
+      VariableExprNode* var2 = variable();
+      IncrDecrStmNode* incr_decr;
+      if (token_.type == decaf::token_type::OpArtInc) {
+        match(decaf::token_type::OpArtInc);
+        incr_decr = new IncrStmNode(var2);
+      } else if (token_.type == decaf::token_type::OpArtDec) {
+        match(decaf::token_type::OpArtDec);
+        incr_decr = new DecrStmNode(var2);
+      } else {
+        error(decaf::token_type::OpArtInc);
+      }
+      match(decaf::token_type::ptRParen);
+      BlockStmNode* block = statement_block();
+      return new ForStmNode(new AssignStmNode(var_new, value), condition,
+                            incr_decr, block);
+    }
+    case (decaf::token_type::kwReturn): {
+      match(decaf::token_type::kwReturn);
+      // optional_expr handeled here
+      ExprNode* opt_expr;
+      if (token_.type != decaf::token_type::ptSemicolon) {
+        opt_expr = expr_or();
+      }
+      match(decaf::token_type::ptSemicolon);
+      return new ReturnStmNode(opt_expr);
+    }
+    case (decaf::token_type::kwBreak): {
+      match(decaf::token_type::kwBreak);
+      match(decaf::token_type::ptSemicolon);
+      return new BreakStmNode();
+    }
+    case (decaf::token_type::kwContinue): {
+      match(decaf::token_type::kwContinue);
+      match(decaf::token_type::ptSemicolon);
+      return new ContinueStmNode();
+    }
+    case (decaf::token_type::ptLBrace): {
+      return statement_block();
+    }
+    case (decaf::token_type::Identifier): {
+      return id_start_stm();
+    }
+    default:
+      break;
+  }
+  return new ReturnStmNode(nullptr);
+}
+
+StmNode* HParser::id_start_stm() {
+  string id = token_.lexeme;
+  match(decaf::token_type::Identifier);
+  switch (token_.type) {
+    case decaf::token_type::ptLParen: {
+      match(decaf::token_type::ptLParen);
+      auto ex_list = expr_list();
+      match(decaf::token_type::ptRParen);
+      match(decaf::token_type::ptSemicolon);
+      return new MethodCallExprStmNode(id, ex_list);
+    }
+    case decaf::token_type::OpAssign: {
+      match(decaf::token_type::OpAssign);
+      auto expr = expr_or();
+      match(decaf::token_type::ptSemicolon);
+      return new AssignStmNode(new VariableExprNode(id), expr);
+    }
+    case decaf::token_type::OpArtInc: {
+      match(decaf::token_type::OpArtInc);
+      match(decaf::token_type::ptSemicolon);
+      return new IncrStmNode(new VariableExprNode(id));
+    }
+    case decaf::token_type::OpArtDec: {
+      match(decaf::token_type::OpArtDec);
+      match(decaf::token_type::ptSemicolon);
+      return new DecrStmNode(new VariableExprNode(id));
+    }
+    default:
+      break;
+  }
```

Comment:
I am not sure what should we return here. Throw an exception? what kind of error should happen in this case?

Reply:
I think we should only call given error function. But I'm not sure if we have to check for errors every time when we have a sequence of match statements. Probably.